src/storage-client/src/sink.rs

```diff
@@ -474,25 +497,34 @@ pub async fn determine_latest_progress_record(
             return Ok(None);
         }
 
-        let mut latest_ts = None;
-        let mut latest_offset = None;
-
-        let progress_key_bytes = progress_key.as_bytes();
+        let mut maybe_upper = None;
+        let mut offset_upper = 0;
         while let Some((key, message, offset)) = get_next_message(progress_client, timeout)? {
-            debug_assert!(offset >= latest_offset.unwrap_or(0));
-            latest_offset = Some(offset);
-
-            let timestamp_opt = if &key == progress_key_bytes {
-                let progress: ProgressRecord = serde_json::from_slice(&message)?;
-                Some(progress.timestamp)
-            } else {
-                None
+            assert!(offset_upper <= offset);
+            offset_upper = offset + 1;
+
+            if &key != progress_key.as_bytes() {
+                continue;
+            }
+
+            let new_upper = match serde_json::from_slice::<ProgressRecord>(&message) {
+                Ok(progress) => Antichain::from(progress.frontier),
+                // If we fail to deserialize we might be reading a legacy progress record
+                Err(_) => {
+                    let legacy: LegacyProgressRecord = serde_json::from_slice(&message)?;
```

Comment:
on error you should concatenate the 2 errors together

Reply:
made it log the invalid bytes instead when both of these fail